bold-linkage/main.py

```diff
@@ -0,0 +1,224 @@
+import json
+import logging
+import os
+import uuid
+from datetime import datetime, timezone
+from typing import Dict, List
+
+import requests
+from kafka import KafkaConsumer, KafkaProducer
+from requests.auth import HTTPBasicAuth
+
+logging.basicConfig(format="%(asctime)s - %(message)s", level=logging.INFO)
+
+ODS_TYPE = "ods:type"
+ODS_ID = "ods:id"
+
+
+def start_kafka() -> None:
+    """
+    Start a kafka listener and process the messages by unpacking the image.
+    When done it will republish the object, so it can be validated and stored by the processing service
+    """
+    consumer = KafkaConsumer(
+        os.environ.get("KAFKA_CONSUMER_TOPIC"),
+        group_id=os.environ.get("KAFKA_CONSUMER_GROUP"),
+        bootstrap_servers=[os.environ.get("KAFKA_CONSUMER_HOST")],
+        value_deserializer=lambda m: json.loads(m.decode("utf-8")),
+        enable_auto_commit=True,
+    )
+    producer = KafkaProducer(
+        bootstrap_servers=[os.environ.get("KAFKA_PRODUCER_HOST")],
+        value_serializer=lambda m: json.dumps(m).encode("utf-8"),
+    )
+    for msg in consumer:
+        try:
+            logging.info("Received message: " + str(msg.value))
+            json_value = msg.value
+            specimen_data = json_value["object"]["digitalSpecimen"]
+            result = run_api_call(specimen_data)
+            mas_job_record = map_to_mas_job_record(
+                specimen_data, result, json_value["jobId"]
+            )
+            send_updated_opends(mas_job_record, producer)
+        except Exception as e:
+            logging.exception(e)
+
+
+def map_to_mas_job_record(
+    specimen_data: Dict, results: List[Dict[str, str]], job_id: str
+) -> Dict:
+    """
+    Map the result of the API call to an annotation
+    :param specimen_data: The JSON value of the Digital Specimen
+    :param results: A list of results that contain the queryString and the BOLD EU process identifier
+    :param job_id: The job ID of the MAS
+    :return: Returns a formatted annotation Record which includes the Job ID
+    """
+    timestamp = timestamp_now()
+    if results is None:
+        annotations = list()
+    else:
+        annotations = list(
+            map(
+                lambda result: map_to_annotation(specimen_data, result, timestamp),
+                results,
+            )
+        )
+    mas_job_record = {"jobId": job_id, "annotations": annotations}
```

Comment:
i know this variable is called mas_job_record all over this repo, but it's actually an annotation_event

mas_job_record = metadata record about the job, includes job id, job state, mas_id, time start/completed, etc
annotation_event = object containing annotations, job id, and batch metadata if applicable. it's what the processor receives.